shim/shim.c

```diff
@@ -0,0 +1,631 @@
+
+	struct option prog_opts[] = {
+		{"container-id", required_argument, 0, 'c'},
+		{"proxy-sock-fd", required_argument, 0, 'p'},
+		{"proxy-io-fd", required_argument, 0, 'o'},
+		{"seq-no", required_argument, 0, 's'},
+		{"err-seq-no", required_argument, 0, 'e'},
+		{"debug", no_argument, 0, 'd'},
+		{"help", no_argument, 0, 'h'},
+		{ 0, 0, 0, 0},
+	};
+
+	while ((c = getopt_long(argc, argv, "c:p:o:s:e:dh", prog_opts, NULL))!= -1) {
+		switch (c) {
+			case 'c':
+				shim.container_id = strdup(optarg);
+				break;
+			case 'p':
+				shim.proxy_sock_fd = (int)parse_numeric_option(optarg);
+				if (shim.proxy_sock_fd < 0) {
+					err_exit("Invalid value for proxy socket fd\n");
+				}
+				break;
+			case 'o':
+				shim.proxy_io_fd = (int)parse_numeric_option(optarg);
+				if (shim.proxy_io_fd < 0) {
+					err_exit("Invalid value for proxy IO fd\n");
+				}
+				break;
+			case 's':
+				shim.io_seq_no = (uint64_t)parse_numeric_option(optarg);
+				if (shim.io_seq_no < 0) {
+					err_exit("Invalid value for I/O seqence no\n");
+				}
+				break;
+			case 'e':
+				shim.err_seq_no = (uint64_t)parse_numeric_option(optarg);
+				if (shim.err_seq_no < 0) {
+					err_exit("Invalid value for error sequence no\n");
+				}
+				break;
+			case 'd':
+				debug = true;
+				break;
+			case 'h':
+				print_usage();
+				exit(EXIT_SUCCESS);
+			default:
+				print_usage();
+				exit(EXIT_FAILURE);
+		}
+	}
+
+	if ( !shim.container_id) {
+		err_exit("Missing container id\n");
+	}
+
+	if ( shim.proxy_sock_fd == -1) {
+		err_exit("Missing proxy socket file descriptor\n");
+	}
+
+	if ( shim.proxy_io_fd == -1) {
+		err_exit("Missing proxy I/O file descriptor\n");
+	}
+
+	if (shim.io_seq_no == 0) {
+		err_exit("Missing I/O sequence number\n");
+	}
+
+	shim_log_init(debug);
+
+	/* Using self pipe trick to handle signals in the main loop, other strategy
+	 * would be to clock signals and use signalfd()/ to handle signals synchronously
+	 */
+	if (pipe(signal_pipe_fd) == -1) {
+		err_exit("Error creating pipe\n");
+	}
+
+	// Add read end of pipe to pollfd list and make it non-bocking
+	add_pollfd(poll_fds, &nfds, signal_pipe_fd[0], POLLIN | POLLPRI);
+	if (! set_fd_nonblocking(signal_pipe_fd[0])) {
+		exit(EXIT_FAILURE);
+	}
+	if (! set_fd_nonblocking(signal_pipe_fd[1])) {
+		exit(EXIT_FAILURE);
+	}
+
+	sigemptyset(&sa.sa_mask);
+	sa.sa_flags = SA_RESTART;           /* Restart interrupted reads()s */
+	sa.sa_handler = signal_handler;
+
+	// Change the default action of all signals that should be forwarded to proxy
+	if (! assign_all_signals(&sa)) {
+		err_exit("sigaction");
+	}
+
+	if ( !set_fd_nonblocking(STDIN_FILENO)) {
+		exit(EXIT_FAILURE);
+	}
+
+	add_pollfd(poll_fds, &nfds, STDIN_FILENO, POLLIN | POLLPRI);
+
+	add_pollfd(poll_fds, &nfds, shim.proxy_io_fd, POLLIN | POLLPRI);
+
+	add_pollfd(poll_fds, &nfds, shim.proxy_sock_fd, POLLIN | POLLPRI);
+
+	/*	0 =>signal_pipe_fd[0]
+		1 =>stdin
+		2 =>proxy_io_fd
+		3 =>sockfd
+	*/
+
+	while (1) {
```

Comment:
Since there is no normal code path to break out of this loop, I think this deserves a comment explaining why.